File: java/mergeSort/mergeSort.java

class Main {

	private static int[] merge(int[] left, int[] right) {
		int[] result = new int[left.length + right.length];
		int leftIndex = 0;
		int rightIndex = 0;

		for (int i = 0; i < result.length; ++i) {
			if (leftIndex < left.length) {
				if (rightIndex < right.length) {
					if (left[leftIndex] < right[rightIndex]) {
						result[i] = left[leftIndex++];
					}
					else {
						result[i] = right[rightIndex++];
					}
				}
				else {
					result[i] = left[leftIndex++];
				}
			}
			else {
				result[i] = right[rightIndex++];
			}
		}
		return result;
	}

	private static int[] mergeSort(int[] array) {
		if (array.length <= 1) {
			return array;
		}

		int m = array.length / 2;

		int[] left = new int[m];
		for (int i = 0; i < left.length; ++i) {
			left[i] = array[i];
		}
		int[] right = new int[array.length-m];
		for (int i = 0; i < right.length; ++i) {
			right[i] = array[m+i];
		}

		return merge(mergeSort(left), mergeSort(right));
	}

	private static void printArray(int[] array) {
		for (int e : array) {
			System.out.println(e);
		}
	}

	public static void main(String[] args) {
		int[] array = {12, 11, 13, 5, 6};
		int[] result = mergeSort(array);
		printArray(result);
	}

}

File: cpp/mergeSort/mergeSort.cpp
#include <iostream>

void merge(int* array, int left, int m, int right) {

	int leftSize = m-left+1;
	int rightSize = right-m;
	int* leftArray = new int[leftSize];
	int* rightArray = new int[rightSize];

	for (int i = 0; i < leftSize; ++i) {
		leftArray[i] = array[left+i];
	}
	for (int i = 0; i < rightSize; ++i) {
		rightArray[i] = array[m+1+i];
	}

	int index = left;
	int leftIndex = 0;
	int rightIndex = 0;
	while (leftIndex < leftSize && rightIndex < rightSize) {
		if (leftArray[leftIndex] < rightArray[rightIndex]) {
			array[index++] = leftArray[leftIndex++];
		}
		else {
			array[index++] = rightArray[rightIndex++];
		}
	}

	while (leftIndex < leftSize) {
		array[index++] = leftArray[leftIndex++];
	}
	while (rightIndex < rightSize) {
		array[index++] = rightArray[rightIndex++];
	}
	delete[] leftArray;
	delete[] rightArray;
}

void mergeSort(int* array, int left, int right) {
	if (left >= right) {
		return;
	}

	int m = (left + right) / 2;
	mergeSort(array, left, m);
	mergeSort(array, m+1, right);
	merge(array, left, m, right);
}

void printArray(int* array, int len) {
	for (int i = 0; i < len; ++i) {
		std::cout << array[i] << " ";
	}
	std::cout << "\n";
}

int main() {
	int* array = new int[]{0, 8, 0, 3, 1, 47, 100, 2, 30, 8};
	mergeSort(array, 0, 9);
	printArray(array, 10);

	delete[] array;
	return 0;
}
